sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.ClientModel.Options;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// TODO.
+/// </summary>
+public class ClientLoggingPolicy : PipelinePolicy
+{
+    private const double RequestTooLongTime = 3.0; // sec
+    private const int DefaultLoggedContentSizeLimit = 4 * 1024;
+    private static ClientModelEventSource? s_eventSource;
+    private const string DefaultEventSourceName = "System.ClientModel";
+
+    private readonly bool _logContent;
+    private readonly int _maxLength;
+    private readonly string? _assemblyName;
+    private readonly string? _clientRequestIdHeaderName;
+    private readonly bool _isLoggingEnabled;
+
+    internal static ClientModelEventSource Singleton { get => s_eventSource ??= ClientModelEventSource.Create(DefaultEventSourceName, Array.Empty<string>()); } //TODO add traits
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="eventSourceName"></param>
+    /// <param name="eventSourceTraits"></param>
+    /// <param name="options"></param>
+    protected ClientLoggingPolicy(string? eventSourceName, string[]? eventSourceTraits = default, DiagnosticsOptions? options = default) : this(options)
```

Comment:
Looking at this now, if we end up using this approach, maybe we would want to generalize the name so it doesn't say event source, since we didn't want hints of that in the public API.  But I know this piece is in flux right now, so we can wait on that piece -- if we move forward with it, it'd be good to figure out if there's a comparable concept in the ILogger world and use a generalized name.  Traits might be a little harder, but we can figure that out too if needed.